storage: minimize retries with 1PC commits

spencerkimball opened the pull request:
Introduce a new `NoRefreshSpans` field to the `EndTransactionRequest`
arguments. This specifies that a serializable isolation transaction has
encountered no refresh spans. On a 1PC commit, this can be used to
avoid serializable restarts by re-executing the 1PC transaction at an
appropriately higher timestamp in the event of the timestamp being
forwarded by the timestamp cache or because of write-too-old errors.
    
When evaluating a write batch, we now allow a local retry for write
too old errors for non-transactional batches, and for serializable
1PC txns where `NoRefreshSpans` is true.
    
Release note: None

spencerkimball committed:
storage: minimize retries with 1PC commits

Introduce a new `NoRefreshSpans` field to the `EndTransactionRequest`
arguments. This specifies that a serializable isolation transaction has
encountered no refresh spans. On a 1PC commit, this can be used to
avoid serializable restarts by re-executing the 1PC transaction at an
appropriately higher timestamp in the event of the timestamp being
forwarded by the timestamp cache or because of write-too-old errors.

When evaluating a write batch, we now allow a local retry for write
too old errors for non-transactional batches, and for serializable
1PC txns where `NoRefreshSpans` is true.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22315)
<!-- Reviewable:end -->

spencerkimball commented:
@nvanbenschoten this now evaluates the full batch for both transactional and non-transactional batches to find the highest timestamp to satisfy write too old errors. Local retries on the replica at evaluation time are allowed for non-transactional batches always and for serializable 1PC transactional batches, as long as they have the `NoRefreshSpans` flag set to true.

nvanbenschoten commented:
Thanks @spencerkimball, I'm more comfortable with the change now that we're attempting to minimize the timestamp that these transactions will be pushed to. This eliminates the increase in `WriteTooOld` errors that you were seeing before, right?

spencerkimball commented:
@nvanbenschoten it does eliminate those yes.

nvanbenschoten commented:
Reviewed 5 of 7 files at r1.
Review status: 5 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/roachpb/api.proto, line 458 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NKjo91inR0OUqWE4O:-L4NKjo91inR0OUqWE4P:b-9i9eyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/roachpb/api.proto#L458)):*
> ```protobuf
>   // intents are left in the event of an error.
>   bool require_1pc = 6 [(gogoproto.customname) = "Require1PC"];
>   // Set to true if this transaction is serializable isolation and has
> ```

Does it ever make sense to set this for snapshot txns?

---

*[pkg/roachpb/api.proto, line 461 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NKuIsBVQ_o3xwatkA:-L4NKuIsBVQ_o3xwatkB:b315iku) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/roachpb/api.proto#L461)):*
>  If true, one phase commits can use the executing server's
>   // hlc clock now timestamp to avoid retries.

replace "use the executing server's hlc clock now timestamp" with "increase their timestamp on the executing server"

---

*[pkg/roachpb/api.proto, line 462 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NLEyq587tfHrvxQSq:-L4NLEyq587tfHrvxQSr:brx4j4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/roachpb/api.proto#L462)):*
> ```protobuf
>   // etc.). If true, one phase commits can use the executing server's
>   // hlc clock now timestamp to avoid retries.
>   bool no_refresh_spans = 7;
> ```

Do you think we should make this flag even more general? Something like `committed_to_timestamp` or `timestamp_negotiable`.

As discussed with @andreimatei, we don't want to increase our transaction timestamp if we've done anything that relies on our initial timestamp. This includes having performed reads and writes at the timestamp but can also include leaking the timestamp in other ways like reporting it to SQL functions.

---

*[pkg/storage/replica.go, line 5209 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NNMmgBaw6YLDRA2w3:-L4NNMmgBaw6YLDRA2w4:b920ivi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/storage/replica.go#L5209)):*
> ```Go
> 
> 	rec := NewReplicaEvalContext(r, spans)
> 	// We can retry locally if this is a non-transactional request.
> ```

nit: pull `ba.Txn == nil` into a variable like you did above with `retryLocally` so that it's easier to see the parallels.

---

*[pkg/storage/replica.go, line 5273 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NOXQDBa7cJCWftfbJ:-L4NOXQDBa7cJCWftfbK:bjwu3cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/storage/replica.go#L5273)):*
> ```Go
> 	if retry, reason := isEndTransactionTriggeringRetryError(ba.Txn); retry {
> 		// We can still attempt the one phase commit if the end
> 		// transaction request specifies there are no refresh spans.
> ```

Is this actually possible? Doesn't `isEndTransactionTriggeringRetryError == true` imply that the txn has performed previous writes?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 5 of 7 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/api.proto, line 458 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NKjo91inR0OUqWE4O:-L4NcBMu741sOd1pcNhc:b-l9cbnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/roachpb/api.proto#L458)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does it ever make sense to set this for snapshot txns?
</blockquote></details>

We don't track refresh spans for snapshot txns because we don't use them for refreshing, so this optimization doesn't apply to SNAPSHOT.

---

*[pkg/roachpb/api.proto, line 461 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NKuIsBVQ_o3xwatkA:-L4NdLAX0e8CJ8gh-Mje:b-14smfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/roachpb/api.proto#L461)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

>  If true, one phase commits can use the executing server's
>   // hlc clock now timestamp to avoid retries.

replace "use the executing server's hlc clock now timestamp" with "increase their timestamp on the executing server"
</blockquote></details>

Oops. this was a leftover bit.

---

*[pkg/roachpb/api.proto, line 462 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NLEyq587tfHrvxQSq:-L4NdenYCwEfJF2Gag3s:b-jhbl7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/roachpb/api.proto#L462)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you think we should make this flag even more general? Something like `committed_to_timestamp` or `timestamp_negotiable`.

As discussed with @andreimatei, we don't want to increase our transaction timestamp if we've done anything that relies on our initial timestamp. This includes having performed reads and writes at the timestamp but can also include leaking the timestamp in other ways like reporting it to SQL functions.
</blockquote></details>

OK I'm fine with that. Changed it to `can_forward_commit_timestamp`. "negotiable" sounds like it could do something other than move forward.

---

*[pkg/storage/replica.go, line 5209 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NNMmgBaw6YLDRA2w3:-L4NfDz0CLcXEJWTWr_9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/storage/replica.go#L5209)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: pull `ba.Txn == nil` into a variable like you did above with `retryLocally` so that it's easier to see the parallels.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 5273 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4NOXQDBa7cJCWftfbJ:-L4Nc0xN1JlcBhhX6zkl:bjeqpqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2a562e39935880bd8a2ad218994c6a260a5d938/pkg/storage/replica.go#L5273)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this actually possible? Doesn't `isEndTransactionTriggeringRetryError == true` imply that the txn has performed previous writes?
</blockquote></details>

Yes, happens all the time if the batch affects keys which have newer timestamps from the timestamp cache. I improved the comment.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@nvanbenschoten I realize we can actually avoid retries on non fast path (non-1PC) transactions as well with this knowledge about having no refresh spans.

tschottdorf commented:
Looks like this should reference (and potentially close)
https://github.com/cockroachdb/cockroach/issues/15797.

On Fri, Feb 2, 2018 at 6:35 PM Spencer Kimball <notifications@github.com>
wrote:

> @nvanbenschoten <https://github.com/nvanbenschoten> I realize we can
> actually avoid retries on non fast path (non-1PC) transactions as well with
> this knowledge about having no refresh spans.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22315#issuecomment-362740926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135GeShxSMu-G0QQYfE6kd32aWxBrUks5tQ5ujgaJpZM4R2cRc>
> .
>
-- 

-- Tobias

spencerkimball commented:
Fixes #15797

With this change, `kv --concurrency=8 --cycle-length=1` exhibits:
```
_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
      1s        0         2855.9         2855.9      2.8      3.4      5.0      6.8
      2s        0         2950.1         2903.0      2.8      3.0      4.7      6.8
      3s        0         2886.7         2897.6      2.8      3.3      5.8      7.9
      4s        0         2899.1         2898.0      2.8      3.1      5.8      7.3
      5s        0         2937.1         2905.8      2.6      3.0      6.3      8.9
      6s        0         2915.8         2907.5      2.8      3.0      6.6      7.1
      7s        0         2862.1         2901.0      2.8      3.1      7.1     11.5
      8s        0         2717.8         2878.1      2.8      3.9      8.9     15.2
      9s        0         2750.0         2863.8      2.8      3.5      8.1     11.5
     10s        0         2872.2         2864.7      2.8      3.1      8.4      9.4

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   10.0s        0          28649         2864.7      2.8      2.8      3.3      6.8     15.2
```

Previous sha's performance with same options has:
```
_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
      1s        0          462.7          462.7     11.5     48.2     62.9     79.7
      2s        0          449.4          456.0     16.3     46.1     65.0     88.1
      3s        0          441.6          451.2     16.8     48.2     58.7     79.7
      4s        0          458.0          452.9     15.7     46.1     52.4     75.5
      5s        0          438.2          450.0     16.8     48.2     60.8     71.3
      6s        0          451.4          450.2     10.0     46.1     56.6     71.3
      7s        0          361.1          437.5     22.0     56.6     75.5    109.1
      8s        0          416.1          434.8      8.9     52.4     75.5     96.5
      9s        0          440.3          435.5     13.6     50.3     58.7     62.9
     10s        0          451.4          437.0      9.4     46.1     58.7     79.7

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   10.0s        0           4371          437.0     18.3     14.2     48.2     65.0    109.1
```

tschottdorf commented:
Nice!

On Fri, Feb 2, 2018, 7:15 PM Spencer Kimball <notifications@github.com>
wrote:

> Fixes #15797 <https://github.com/cockroachdb/cockroach/issues/15797>
>
> With this change, kv --concurrency=8 --cycle-length=1 exhibits:
>
> _elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
>       1s        0         2855.9         2855.9      2.8      3.4      5.0      6.8
>       2s        0         2950.1         2903.0      2.8      3.0      4.7      6.8
>       3s        0         2886.7         2897.6      2.8      3.3      5.8      7.9
>       4s        0         2899.1         2898.0      2.8      3.1      5.8      7.3
>       5s        0         2937.1         2905.8      2.6      3.0      6.3      8.9
>       6s        0         2915.8         2907.5      2.8      3.0      6.6      7.1
>       7s        0         2862.1         2901.0      2.8      3.1      7.1     11.5
>       8s        0         2717.8         2878.1      2.8      3.9      8.9     15.2
>       9s        0         2750.0         2863.8      2.8      3.5      8.1     11.5
>      10s        0         2872.2         2864.7      2.8      3.1      8.4      9.4
>
> _elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
>    10.0s        0          28649         2864.7      2.8      2.8      3.3      6.8     15.2
>
> Previous sha's performance with same options has:
>
> _elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
>       1s        0          462.7          462.7     11.5     48.2     62.9     79.7
>       2s        0          449.4          456.0     16.3     46.1     65.0     88.1
>       3s        0          441.6          451.2     16.8     48.2     58.7     79.7
>       4s        0          458.0          452.9     15.7     46.1     52.4     75.5
>       5s        0          438.2          450.0     16.8     48.2     60.8     71.3
>       6s        0          451.4          450.2     10.0     46.1     56.6     71.3
>       7s        0          361.1          437.5     22.0     56.6     75.5    109.1
>       8s        0          416.1          434.8      8.9     52.4     75.5     96.5
>       9s        0          440.3          435.5     13.6     50.3     58.7     62.9
>      10s        0          451.4          437.0      9.4     46.1     58.7     79.7
>
> _elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
>    10.0s        0           4371          437.0     18.3     14.2     48.2     65.0    109.1
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22315#issuecomment-362748626>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135F1R0cL33Zg1owPIaH9YoVlwSra5ks5tQ6U2gaJpZM4R2cRc>
> .
>


-- 

-- Tobias

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> that's a nice improvement!

> I realize we can actually avoid retries on non fast path (non-1PC) transactions as well with this knowledge about having no refresh spans.

Yeah, that's a good point. We can always allow transaction timestamps to be pushed forward to avoid retries up until the timestamp is leaked for the first time. To make that change I think we'll need to move the `CanForwardCommitTimestamp` flag from `EndTransactionRequest` onto the `Transaction` proto itself.

---

Reviewed 4 of 6 files at r2.
Review status: 6 of 7 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/roachpb/data.go, line 1145 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4ToSUr6svHslS3aTRE:-L4ToSUr6svHslS3aTRF:b-hxtbc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/data.go#L1145)):*
> ```Go
> 	}
> 	// Also forward by the existing timestamp.
> 	ts.Forward(err.ExistingTimestamp.Add(0, 1))
> ```

Why was this suddenly caught? Did part of this change cause it to become more important?

`s/.Add(0, 1)/.Next()/`

Also, It's unfortunate that the discrepancy between `ReadWithinUncertaintyIntervalError.ExistingTimestamp` and `WriteTooOldError.ActualTimestamp` caused us to make this mistake. I'd prefer we change them to both reflect either the `existing value's timestamp` or the `existing value's timestamp+1`. Either way, let's prevent future confusion. If this isn't easily possible because of the migration concern, let's at least be more explicit about the disagreement.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4Tqerh2XXGcJIArqlo:b-olborb)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@spencerkimball, do you mind trying this change out on TPC-C as well? https://github.com/cockroachdb/cockroach/issues/15797#issuecomment-330998850 implies that TPC-C might also be affected by an approach like this.

bdarnell commented:
Reviewed 1 of 7 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 464 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-L4Xw2vdHR5DHZrl2LYW:-L4Xw2ve1P_xturtLCK1:b9qwm8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L464)):*
> ```Go
> 			tc.mu.intentsSizeBytes = intentsSizeBytes
> 
> 			if tc.mu.meta.Txn.IsSerializable() && tc.mu.meta.RefreshValid &&
> ```

While this is all fresh in your mind, make sure there is commentary somewhere about why we can't do this for snapshot transactions. (is it just because we don't accumulate refresh spans for non-serializable transactions or is there a deeper reason? If we weren't thinking about getting rid of snapshot completely, it might be worth saving refresh spans just to enable this optimization)

Maybe setting the isolation to snapshot should set RefreshValid to false so we don't need the separate IsSerializable check here. 

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4XwdnICtz6kNCfWrV4:b-93x81i) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
> ```Go
> 			if tc.mu.meta.Txn.IsSerializable() && tc.mu.meta.RefreshValid &&
> 				len(tc.mu.meta.RefreshReads) == 0 && len(tc.mu.meta.RefreshWrites) == 0 {
> 				et.CanForwardCommitTimestamp = true
> ```

Can the DistSender split the batch below this point? We need to make sure to clear the CanForwardCommitTimestamp flag if that happens. 

---

*[pkg/storage/replica.go, line 5238 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-L4YAOwd8cKaowGDMBXy:-L4YAOwd8cKaowGDMBXz:b1cquoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/storage/replica.go#L5238)):*
> ```Go
> 		br, res, pErr = evaluateBatch(ctx, idKey, batch, rec, ms, ba)
> 		// If we can retry, set a higher batch timestamp and continue.
> 		if wtoErr, ok := pErr.GetDetail().(*roachpb.WriteTooOldError); ok && canRetry {
> ```

If `retries` is already 1, this will mutate `ba.Timestamp` but won't actually retry, which seems odd. On the other hand, since any concurrent writes should be blocked in the command queue and we look at all requests in the batch, we really shouldn't see more than one WriteTooOldError here. Maybe we should panic if we get more than one?

---

*[pkg/storage/replica.go, line 5487 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-L4YEiVc9zKHg-JEOa6z:-L4YEiVc9zKHg-JEOa7-:b-smiujc) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/storage/replica.go#L5487)):*
> ```Go
> 					}
> 				}
> 				// Return error early on some requests for serializable isolation.
> ```

I think some comments about how we define "returning early" here (i.e. the end of the batch) would be helpful. This is a heuristic that may not always be optimal. Finishing the batch helps us choose the right timestamp in the event that there are multiple conflicts *and* the client retry will end up writing to the same keys the next time around. If the client behavior changes on retries then laying down the intents on this attempt was wasted work. If running the rest of the batch is beneficial, then the same logic might argue for running the entire remainder of the transaction. Returning a WriteTooOldError at the end of the batch is simply choosing a local minimum in the cost/benefit curve, since the cost of continuing the transaction is much less when we're not requiring additional round trips. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@nvanbenschoten are there instructions for running TPC-C?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 464 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4Xw2vdHR5DHZrl2LYW:-L4YiUp6CWNSbBBbF6Mb:b-si060x) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L464)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

While this is all fresh in your mind, make sure there is commentary somewhere about why we can't do this for snapshot transactions. (is it just because we don't accumulate refresh spans for non-serializable transactions or is there a deeper reason? If we weren't thinking about getting rid of snapshot completely, it might be worth saving refresh spans just to enable this optimization)

Maybe setting the isolation to snapshot should set RefreshValid to false so we don't need the separate IsSerializable check here. 
</blockquote></details>

We could do this for snapshot transactions, but I don't think we should. As you point out, we don't accumulate the refresh spans for SNAPSHOT. Not sure it makes sense to add any commentary considering the reason is so straightforward.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4YjnR93GBU7xwoEIu-:bqx977p) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can the DistSender split the batch below this point? We need to make sure to clear the CanForwardCommitTimestamp flag if that happens. 
</blockquote></details>

It can. Originally the name of this flag was `NoRefreshSpans` which was specifically meant to still be accurate for that case, but then it was suggested to change the name and I didn't remember this case when agreeing. I think I'm going to change this back to `NoRefreshSpans`.

---

*[pkg/roachpb/data.go, line 1145 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4ToSUr6svHslS3aTRE:-L4YmE7KCyMa30i-U6KN:b16jhlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/data.go#L1145)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why was this suddenly caught? Did part of this change cause it to become more important?

`s/.Add(0, 1)/.Next()/`

Also, It's unfortunate that the discrepancy between `ReadWithinUncertaintyIntervalError.ExistingTimestamp` and `WriteTooOldError.ActualTimestamp` caused us to make this mistake. I'd prefer we change them to both reflect either the `existing value's timestamp` or the `existing value's timestamp+1`. Either way, let's prevent future confusion. If this isn't easily possible because of the migration concern, let's at least be more explicit about the disagreement.
</blockquote></details>

I added this addition for some unittest cases. It never fires in real life. But turns out that when I first added it, just using `ExistingTimestamp` was incorrect – need to add a logical tick to make sure we don't get another `WriteTooOldError`. Nothing prompted this fix except realizing the original change was incorrect.

---

*[pkg/storage/replica.go, line 5238 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4YAOwd8cKaowGDMBXy:-L4Ymbl7Axd949h9O5EZ:b-id1ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/storage/replica.go#L5238)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If `retries` is already 1, this will mutate `ba.Timestamp` but won't actually retry, which seems odd. On the other hand, since any concurrent writes should be blocked in the command queue and we look at all requests in the batch, we really shouldn't see more than one WriteTooOldError here. Maybe we should panic if we get more than one?
</blockquote></details>

You can get more than one `WriteTooOldError` in the case where a non-transactional batch updates the same key on more than one request. Obviously this can also by extension happen to a transactional batch which is attempting to execute on the fast path. Added a comment.

---

*[pkg/storage/replica.go, line 5487 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4YEiVc9zKHg-JEOa6z:-L4Ypw_u4QFHDauCH7o-:b50756k) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/storage/replica.go#L5487)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think some comments about how we define "returning early" here (i.e. the end of the batch) would be helpful. This is a heuristic that may not always be optimal. Finishing the batch helps us choose the right timestamp in the event that there are multiple conflicts *and* the client retry will end up writing to the same keys the next time around. If the client behavior changes on retries then laying down the intents on this attempt was wasted work. If running the rest of the batch is beneficial, then the same logic might argue for running the entire remainder of the transaction. Returning a WriteTooOldError at the end of the batch is simply choosing a local minimum in the cost/benefit curve, since the cost of continuing the transaction is much less when we're not requiring additional round trips. 
</blockquote></details>

Added to the comment.

An important part of the story here is when this returning early situation is triggered by a conditional put (increment is not even really a factor in practice). On a conditional put, the chance that the retry will succeed is pretty slim, and so laying down intents is almost always going to be a bad idea. The reason cputs are likely to fail is that they're often either based on a read of the underlying value which will change on the forwarded retry timestamp, or they're expecting an empty value, which won't be true on the retry.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4YiPaR-4IyochvA_mz:b-ywpwiw)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@spencerkimball yes, the load generator is [here](https://github.com/cockroachdb/loadgen/tree/master/tpcc). You should just be able to run `./tpcc -drop -load -warehouses=1`. Try with both `-no-wait=false` and `-no-wait=true`.

---

Review status: 1 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4azbC-5reDWyvGqm_T:b-9w725m)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
FWIW `no-wait=false` won't be very interesting with just 1 warehouse, because the load generator will mostly be sleeping.

spencerkimball commented:
The results are not appreciably different for the `tpcc` load generator with this change.

bdarnell commented:
Review status: 1 of 7 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4cDHs7AQRZBVjYhKb3:b-eqt0tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
> It can. 

Then don't we need to clear this flag (whatever it's called) when DistSender splits the batch? The sub-batches could create refresh spans but the EndTransaction sent on its own would still have NoRefreshSpans set. 

---

*[pkg/roachpb/api.proto, line 463 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4cD6k85yg7kQttxOIe:b-f1muqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
> ```protobuf
>   // path. For this to be true, the transaction must not have
>   // accumulated any refreshable spans (e.g. from reads, cputs, etc.),
>   // and must not have "leaked" a commit timestamp to the SQL client.
> ```

What happened to the part about leaking a commit timestamp to the SQL client? Don't we still need to do that? 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/api.proto, line 463 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4flggj5efQp4mKasXz:b7hqsn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What happened to the part about leaking a commit timestamp to the SQL client? Don't we still need to do that? 
</blockquote></details>

On a related note, I want to check that we're good on other txns that care about their timestamps particularly - the incremental backup roll-up txns. Are those unaffected by this series of changes?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/api.proto, line 463 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4fmTM7BzCSZLWhsrTt:btog5av) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

On a related note, I want to check that we're good on other txns that care about their timestamps particularly - the incremental backup roll-up txns. Are those unaffected by this series of changes?
</blockquote></details>

Cc @danhhz @knz 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/api.proto, line 463 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4fodn27eIuDmFOYwAa:b-w8sfyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Cc @danhhz @knz 
</blockquote></details>

I haven't fully digested this PR, but backups don't use the normal transaction mechanism. They just pick a timestamp (usually more than 10s in the past) and keep trying the export kv request until it goes through. Incremental backups work the same way, just with an additional param in the request to indicate the beginning of the time span we care about

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

knz commented:
Here's an example 1PC query that would leak a timestamp to the user:

`insert into foo(pk) values (cluster_logical_timestamp()) returning pk`

knz commented:
@andreimatei some of my offline comments from yesterday are not applicable here because the optimization only applies to 1PC transactions. Our consistency checks are mostly 2PC

spencerkimball commented:
@knz, is that query supposed to guarantee the **commit** timestamp, or just a timestamp from the cluster?

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4fqeMW8jA3Uyk1zW14:b5h4432) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> It can. 

Then don't we need to clear this flag (whatever it's called) when DistSender splits the batch? The sub-batches could create refresh spans but the EndTransaction sent on its own would still have NoRefreshSpans set. 
</blockquote></details>

We don't need to clear the flag right now and in fact we still want it. Even when the batch is split, if `NoRefreshSpans` is true, it means we can commit a serializable txn despite the timestamp timestamp having moved forward. I'm going to make a further change to effect that optimization. When we do that, we'll need to reset this flag more precisely at the dist sender to take into consideration any part of the split batch, already executed, which will generate a refresh span (e.g. cputs and incs).

---

*[pkg/roachpb/api.proto, line 463 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4ft1MTABhzgxhRgxD5:b-s7u0pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I haven't fully digested this PR, but backups don't use the normal transaction mechanism. They just pick a timestamp (usually more than 10s in the past) and keep trying the export kv request until it goes through. Incremental backups work the same way, just with an additional param in the request to indicate the beginning of the time span we care about
</blockquote></details>

We don't currently leak the commit timestamp to the SQL client before the commit (there is some testing case which @andreimatei mentioned). If we ever do, I would suggest an additional flag to indicate that, as opposed to trying to have the txn coord sender set a policy for storage, when there are layers in between which may affect the viability of that policy.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4fu2Ig0q9ncmoxiu7o:b-6ykaur)*
<!-- Sent from Reviewable.io -->

knz commented:
>  is that query supposed to guarantee the commit timestamp, or just a timestamp from the cluster?

It's guaranteed to provide a timestamp that is monotonically increasing in transaction commit order.

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4fvyifB0AlDiGWElsz:b1hsq6c)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@knz, what does that mean? Could you elaborate? We definitely can't provide a monotonically increasing timestamp that corresponds to global commit order without providing linearizable semantics.

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4fziSc52UvkNONPSfX:b-m92dxq)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Within a session, this value will always be monotonically increasing on successive transaction commits.

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4g--jf6oBDd5pePW1c:b-8c44rn)*
<!-- Sent from Reviewable.io -->

knz commented:
>  Could you elaborate? We definitely can't provide a monotonically increasing timestamp that corresponds to global commit order without providing linearizable semantics.

yes the monotonicity is only required across transactions that have touching keys (they carry causality through the db)

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4g8WB79VBQeeNjzCOT:b-xqt4sx)*
<!-- Sent from Reviewable.io -->

knz commented:
Also just to clarify I do not want to mischaracterize the seriousness of the problem: this feature `cluster_logical_timestamp()` is specific to cockroachdb and I don't recally we've been very clear about specifying its semantics. Perhaps it does not matter what your patch does to it.

*However* I am still thinking about the consequences of the patch to `now()` (also called `transaction_timestamp()`) which has very clear semantics in pg's dialect - it also must be monotonic in commit order, and I don't know how this patch impacts it.

spencerkimball commented:
This is going to be a problem with these changes I've made. We must disable txn coord sender and replica retries in the event that SQL has leaked a timestamp which must equal the commit timestamp. I'm going to need help here to get that signal out of the executor and passed to the `client.Txn`.

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4gAHse3ujmFhUlLm1A:bk5ohib)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> However I am still thinking about the consequences of the patch to now() (also called transaction_timestamp()) which has very clear semantics in pg's dialect - it also must be monotonic in commit order, and I don't know how this patch impacts it.

I'm not clear what it means for `now()` to be monotonic in commit order. `now()` is tied to the transaction start time, not the commit time. Isn't it?

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4gAytMAU_NeNifazSi:b-p7ckvz)*
<!-- Sent from Reviewable.io -->

knz commented:
> now() is tied to the transaction start time, not the commit time. Isn't it?

It used to be that the txn start time was also the txn commit time, so it worked that way yes. Now is not the case any more.

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4gBS6y4AZA-pLeHf2j:b-8w5rf1)*
<!-- Sent from Reviewable.io -->

knz commented:
@bdarnell I checked and our jepsen tests do not exercise the 1PC code path. Should we schedule a roadmap item to make it happen in 2.1?

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4gBdwOCntOEn1UkV6d:bl6qucg)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
The txn start time was not always the txn commit time, as with SNAPSHOT txns. So `now()` if it used to "work", didn't actually work with both isolation levels.

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4gBxgq5XF6qFcQ7keB:b-3amnmk)*
<!-- Sent from Reviewable.io -->

knz commented:
yeah I am not concerned about  snapshot. We never pretended to do anything correct with it anyway :) 
The jepsen test suite didn't exercise snapshot either (AFAIK)

Meanwhile, I *think* we can defer working on this until 2.1 work starts, assuming we clearly document what is going on. A user who wishes to preserve the pre-optimization semantics can tweak their SQL code to force the txn to become 2PC, in which case I believe the previous behavior holds (is that correct?)

petermattis commented:
> It used to be that the txn start time was also the txn commit time, so it worked that way yes. Now is not the case any more.

I must have missed context here: why does `now()` need to be the transaction commit time? You're clearly worried about something that I'm not understanding. 

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-:-L4gBz8PEy2jra1__Yjc:bt6a2sn)*
<!-- Sent from Reviewable.io -->

knz commented:
test monotonic inserts: `insert into t(x, y) select max(x)+1, now()` from multiple concurrent clients must see the two columns `x` and `y` grow monotonically. This is only true if `now()` is monotonic with commit order.

There are variants of this that use multiple tables that check this is also true even when the txns do not conflict on writes (just on reads).

knz commented:
That monotonic insert test as written above however does not use 1PC (because it needs to read the max) so it's not impacted. But I wonder if there are other similar tests that do use 1PC and that would be impacted.

andreimatei commented:
This particular PR is about 1PC, but the broader set of changes affects all txns.

knz commented:
And just to be clear, a disclaimer: please take my comments above with a grain of salt, it's been a while since I last checked what these tests were actually testing. I recall we went through multiple iterations back then to properly define what tests were correct (i.e. the behavior they were testing was strictly required from the most general definition of serializable isolation).

petermattis commented:
> test monotonic inserts: insert into t(x, y) select max(x)+1, now() from multiple concurrent clients must see the two columns x and y grow monotonically. This is only true if now() is monotonic with commit order.
>
> There are variants of this that use multiple tables that check this is also true even when the txns do not conflict on writes (just on reads).

Ok, this is vastly different than saying `now() must be monotonic with commit order` because there are other interactions between the two transactions which provide that guarantee. Without those interactions, I don't see how or why we would provide that guarantee about `now()`.

knz commented:
Yes that is why above I said "only true if the txns have touching keys"

bdarnell commented:
> I checked and our jepsen tests do not exercise the 1PC code path. Should we schedule a roadmap item to make it happen in 2.1?

It might be nice, but I wouldn't prioritize it. The current issue about commit timestamps changing applies to both one- and two-phase transactions. 

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4gMjobwFce8k83LCVV:b-ubdxsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

We don't need to clear the flag right now and in fact we still want it. Even when the batch is split, if `NoRefreshSpans` is true, it means we can commit a serializable txn despite the timestamp timestamp having moved forward. I'm going to make a further change to effect that optimization. When we do that, we'll need to reset this flag more precisely at the dist sender to take into consideration any part of the split batch, already executed, which will generate a refresh span (e.g. cputs and incs).
</blockquote></details>

I don't understand. If the batch is split, then the first parts of the batch could create refresh spans, invalidating the NoRefreshSpans flag on the EndTransactionRequest. (I'm talking about the splitting by type of `BatchRequest.Split`, not the splitting by key of `kv.truncate`)

---

*[pkg/roachpb/api.proto, line 463 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4gNrhjK5fndt1U4Pej:bivanbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
> We don't currently leak the commit timestamp to the SQL client before the commit 

But unless you're in snapshot isolation, if there's a change in the timestamp the transaction would restart and the SQL layer would see the updated OrigTimestamp. 

I'm not too concerned personally about most of the SQL timestamp functions (although I remember there was a lot of consideration given to what we should provide here, so maybe I'm forgetting something important). However, `cluster_logical_timestamp` does need to see the commit timestamp. I'm pretty sure this could break the `monotonic` jepsen test. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22315#-:-L4gRkl6Gdwsjzrq97_7:bkznx86)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4lAoFHAYuQ_xuzOUIU:b-m6fjt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't understand. If the batch is split, then the first parts of the batch could create refresh spans, invalidating the NoRefreshSpans flag on the EndTransactionRequest. (I'm talking about the splitting by type of `BatchRequest.Split`, not the splitting by key of `kv.truncate`)
</blockquote></details>

Well the code in this particular PR only executes on the 1PC path, so the batch could not have been split. The change I'm working on now will allow replica-side retry optimization, and that change resets the `NoRefreshSpans` flag if the batch is split and if the previously-executed split of the batch had refresh spans.

---

*[pkg/roachpb/api.proto, line 463 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4cD6k78ciy8kSzLwEb:-L4lBHIy6tscRkvby3b5:b-9tnnqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/roachpb/api.proto#L463)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> We don't currently leak the commit timestamp to the SQL client before the commit 

But unless you're in snapshot isolation, if there's a change in the timestamp the transaction would restart and the SQL layer would see the updated OrigTimestamp. 

I'm not too concerned personally about most of the SQL timestamp functions (although I remember there was a lot of consideration given to what we should provide here, so maybe I'm forgetting something important). However, `cluster_logical_timestamp` does need to see the commit timestamp. I'm pretty sure this could break the `monotonic` jepsen test. 
</blockquote></details>

Looks like we're having to handle this case after all. The fixes are in an upcoming change.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4lmqv28LbWqidCqhHM:bmfpmle) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
> Well the code in this particular PR only executes on the 1PC path

Ah, that's what I was missing.

> The change I'm working on now will allow replica-side retry optimization, and that change resets the NoRefreshSpans flag if the batch is split and if the previously-executed split of the batch had refresh spans.

OK. That change is going to cause problems for mixed-version clusters, since nodes that only have this commit will be incorrectly sending NoRefreshSpans on EndTransactionRequests that have been split. An easy way to get around that would be to change the tag number for the NoRefreshSpans proto field. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4lovqs0LTviB2xpsJj:b3uyvri) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> Well the code in this particular PR only executes on the 1PC path

Ah, that's what I was missing.

> The change I'm working on now will allow replica-side retry optimization, and that change resets the NoRefreshSpans flag if the batch is split and if the previously-executed split of the batch had refresh spans.

OK. That change is going to cause problems for mixed-version clusters, since nodes that only have this commit will be incorrectly sending NoRefreshSpans on EndTransactionRequests that have been split. An easy way to get around that would be to change the tag number for the NoRefreshSpans proto field. 

</blockquote></details>

I didn't realize we needed to handle mixed-version clusters on these unstable point releases. I don't think that should be a requirement, especially for a problem like this.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 466 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315#-L4XwdnICtz6kNCfWrV3:-L4lpGnqDUPMYT03oqLP:buna1w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/662e706a3734286ad26c44917b03e3c6a0211488/pkg/kv/txn_coord_sender.go#L466)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I didn't realize we needed to handle mixed-version clusters on these unstable point releases. I don't think that should be a requirement, especially for a problem like this.
</blockquote></details>

We do rolling upgrades of our continuous and test clusters, and sometimes tell customers to do the same. We generally don't do this with critical data, but it's still good to get into the habit of compatibility between versions at any time. (In this case, the best solution would have been to clear the NoRefreshSpans on split in the original commit, but now that it's there and not cleared I think we should change the tag number to preserve compatibility)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22315)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented on pkg/storage/replica.go:
```diff
@@ -1995,7 +1995,13 @@ func (r *Replica) updateTimestampCache(
+	// as when the request is retried locally on WriteTooOldErrors.
 	ts := ba.Timestamp
+	if br != nil {
+		ts = br.Timestamp
```
@spencerkimball or @bdarnell, do you know what makes this safe in relation to lease transfers? Specifically, a batch goes through the following steps:
```
1. update the local node's clock with `ba.Timestamp` when passing through the Store
2. under lock, we check whether a lease transfer is in flight
2a. if so, wait for the lease to stabilize
3. if not, evaluate the batch without ever coordinating with possible lease transfers again
4. bump timestamp cache with batch's timestamp (this code)
```

From the lease transfers point of view:
```
1. lock
2. grab a local clock reading
3. start transfer with that time, which will be used as the new leaseholder's tscache lower bound
4. unlock
```

Since this timestamp is larger than the original `ba.Timestamp`, isn't it possible that it's contribution to the timestamp cache is missed? Are we relying on some implicit assumption that this timestamp will never be bumped past `clock.Now()`? Is that a fair assumption to make in all cases? Are we assuming that this timestamp will always be based on existing data on the Range, which in turn must be at a lower timestamp than the clock on the Store that houses it? If so, I don't see how we're confident about that conclussion, given that we only set transactional batches' timestamps [to the txn's orig timestamp](https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/batch.go#L41) (where it reads) and we don't include a timestamp when resolving intents.

bdarnell replied:
Yes, I think we're implicitly relying on an assumption that any KV data that might bump the response timestamp will have been through some path that would have updated the HLC. I don't know whether that's always valid. It sounds like we might be missing some cases to pull timestamps out of the request (like in intent resolution).

spencerkimball replied:
It's safe because the `br.Timestamp` is only ever forwarded to the timestamp of a key written to the current range. In other words, the local clock must already be at or past any value of key written to the range. You might want to augment the comment here.

nvanbenschoten replied:
This thought experiment blossomed into the beautiful issue we see in https://github.com/cockroachdb/cockroach/issues/36431. Over there we note that even if we did update the leaseholder's HLC clock on intent resolution, we'd still run into issues with transaction uncertainty when intents are moved forward. 

We'll need to address that issue by enforcing clock updates during txn refreshes. In addition, we should do more to ensure that clock updates are never missed, even when `BatchRequests` don't explicitly set a header timestamp. Trying to address this on a case-by-case basis (e.g. https://github.com/cockroachdb/cockroach/pull/35297 and https://github.com/cockroachdb/cockroach/pull/36307) seems like a fraught effort. There has been some discussion about adding a gRPC interceptor to deal with clock updates at a lower-level than in the `BatchRequest` domain.
